### ISSUE TITLE
Implement ECC support for Curve P-256, P-384, P-521 (RFC6637)

### ISSUE THREAD

**Issue by tanx** [openpgpjs/openpgpjs#427]
See http://tools.ietf.org/html/rfc6637

This issue tracks the implementation of ECDSA and ECDH for "Curve P-256", "Curve P-384", and "Curve P-521". The advantage of these curves is that they have already been standardized for OpenPGP in RFC 6637 and are supported natively by both [WebCrypto](https://www.chromium.org/blink/webcrypto) and [node.js crypto](https://nodejs.org/api/crypto.html).

**Comment by fpietrosanti** [comment]
I feel that those should be given precedence compared to Ed25519 of #428 because of wider WebCrypto support and together with https://github.com/openpgpjs/openpgpjs/issues/427 could lead to a fully native implementation, that's imho a major benefit.

**Comment by fpietrosanti** [comment]
Any progress or effort estimation regarding this very specific implementation?

**Comment by tanx** [comment]
> Any progress or effort estimation regarding this very specific implementation?

Nope. OTF has not responded to my application to the [Core Infrastructure Fund](https://www.opentech.fund/requests/core-infrastructure-fund).

I'd be up to implement this though if anyone has budget for it.

**Comment by prusnak** [comment]
What budget are we talking about here? We could do a crowdfunding via Bitcoin for this.

**Comment by tanx** [comment]
> What budget are we talking about here?

For ECC support including NIST Curves and Ed25119, I'd estimate 15-20k EUR. I'm usually on the pessimistic side when it comes to estimating cost though. But whatever is left, could be used to fund other features on the [v3.0 roadmap](https://github.com/openpgpjs/openpgpjs/milestones/v3.0).

> We could do a crowdfunding via Bitcoin for this. I am willing to manage it, if you don't have time for administration.

It would be great if you could manage a crowdfunding!

**Comment by evilaliv3** [comment]
Interesting patch @ismaelbej

ive to surely find the time to test a bit this implementation and to evaluate the possible integration.

could you please elaborate on the status of the implementation and on the possible issues if present?
the code is all written added is all authored by you or is copied from other libraries?
i see that the unit tests are not passing; are you able to make them work before we start digging into the review?

thank you so much!

**Comment by ismaelbej** [comment]
Hi @evilaliv3 

This is a follow up to a previous submit in #26 for OpenPGPjs v1 that mainly supported secp256k1, it is rebased agains v2, ensure it works with p256, p384 and p521, add some more tests and refactored to be easier to add other backend like webcrypto. 

The code was written following RFC 6637 (read and write packets), and RFC 3394 (key wrapping), using certificates created with GnuPG v2.1.8 + libgcrypt v1.7.0-beta262.

Fortunatelly the changes are minimal when dealing with the new packets, but it had take some time to understand how OpenPGP works internally. Also the change to use Uint8Array instead of string was a chunk of job when rebasing against v2. 

For ECC the bulk of the job was already done in Indutny's Elliptic so we used it. For v2 we have refactored to make it easier to use others backends like node's crypto o webcrypto. Following the discussion in #26 we researched how to use node's crypto but we stopped when having to wrap everything around certificates.

Another important part of the work was RFC 3394 (key wrapping), at the time we were not aware of aes-kw so we implemented our own. There was no suitable AES decrypt function in the source code, so used AES library.

I've added two new dependencies:

- It uses Indutny's Elliptic for curve cryptography, only added wrappers to make easier to switch backends. Which itself depends on bn.js and hash.js.
- It uses a AES library to implement key wrapping, but should be easy to drop with a suitable replacement.

Possible issues:

- RFC 6637 add some fields that are not MPI to the public and private key packes, and the currently the OpenPGP implementation assumes that all fields are MPI. So there are some ugly code to work around that.
- To create keys the parameters passed from upper layers are not enough to select different curves. I've added an extra parameter 'curve'. And I think there are some workarounds because unit tests do not expect the new parameter.
- RFC recommends to restrict some hash and encryption algorithms depending on the length of the curve. Like p521 should not be used with sha1, but sha256 instead. That is not implemented. GnuPG enforce that rules when importing certificates, but switching the default hash to sha256 works.
- Add another big number library 'bn.js' as dependency.

I'll take a look at the unit tests during the week end, it seems like some of the unit tests do not like the 'const' keyword.

**Comment by evilaliv3** [comment]
Thank you so much for documenting this all.

The code maturity seems really great and is as soon that i can i will give you feedback on the commits.

Regarding the issues that you hilighted:
- For what relates to the MPI we will see togeter, if for now this ugly code is a few and is controlled it can stay. If not we will work together on a clean fix.
- For the curve argument if you can show me where are the workaround i will look also for this on smoothing the interfaces.
- For whar relates the hashing suggested by the RFC what is the reason of not enforcing them?
- For what relates the addition of a new duplicated big number dependency (bn.js) i see no particular issues because on one hand bn.js is a specialized library required by elliptic, from the other side our library is a minimalistic one used few times so the overhead of including both does not exists and we already have and will have to deal with this situations. 

@indutny, by the chance do you think that would be possible to have bs.jn exposed by elliptic so to reuse bn.js in openpgp.js without including it directly? i think anyhow that this could be good also for other adopters of elliptic.

**Comment by tanx** [comment]
All versions of node.js up until (and including) v4.x require strict mode for ES6 features. This was removed in the current version of V8 that ships with node v6.x and Chrome stable.


For the openpgp.js build should not be an issue as Bable transpiles all ES6 modules to CommonJS and prepends `'use strict';`. Since there is no build step for the unit tests, we need to write them in ES5 to be compatible in all of our target environments on travis/saucelabs.

**Comment by indutny** [comment]
@evilaliv3 it should not appear twice in the final distribution if the versions of bn.js are the same (which is pretty easy to keep up with `^...` version specification in package.json)

**Comment by evilaliv3** [comment]
thanks @indutny i will investigate this, because what i'm thinking is that if elliptic embeds bn.js and we embed it also it will always embed twice because when one use elliptic elliptic comes already with bn builtin. or am i wrong?

@tanx probably i was not clear in what i was saying.
nowadays we have various modules that in an explicit way ask for use both the ```use strict```  ```const``` but that are not facing the issue that you can see on travis for this pull request.
how is this possible?

**Comment by indutny** [comment]
@evilaliv3 it should share the same instance/installation of bn.js between the modules.

**Comment by ismaelbej** [comment]
@evilaliv3 I'll annotate in the pull request the parts about MPI that I consider ugly. Will do the same about the extra parameters 'curve' and 'material'.

Restrict the hash and encryption algorithms depending on the curve/key type can be done. I just have not found enough time for properly fix it.

**Comment by WilliamSWoodruff** [comment]
Is there an update on this? I'm looking for a lib to use in a browser extension to work with my Node CLI that uses gpg2

**Comment by bartbutler** [comment]
No update that I'm aware of, though hopefully will have the bandwidth to focus on this soon, given that it's mostly done so the reward/investment ratio is high.

**Comment by NoSubstitute** [comment]
Any hope for progress soon?

**Comment by wswoodruff** [comment]
Also still interested

**Comment by bartbutler** [comment]
It is being actively developed, stay tuned!

**Comment by mahrud** [comment]
Merged! :dancer:
